### PR TITLE
Support on azure-pipelines.yml

### PR DESCRIPTION
**Pull request by tonybaloney**
Starting a WIP PR just so I can see the build status

### CODE REVIEW

**Comment by tonybaloney** [comment]
@hynek once you change https://github.com/python-attrs/attrs/pull/511/files#diff-fec826feae04e51c0d94076385408bdcR42 with the actual token this build _should_ work :-)

**Comment by tonybaloney** [comment]
Also, you might want to make `pytest-azurepipelines` optional in tox.ini, since the regular user doesn't need it

**Comment by hynek** [comment]
> @hynek once you change https://github.com/python-attrs/attrs/pull/511/files#diff-fec826feae04e51c0d94076385408bdcR42 with the actual token this build _should_ work :-)

I presume you mean the CODECOV_TOKEN? that's a secret variable in Azure Pipelines; is it not available if someone else runs it? I guess that makes sense, since you could open a PR that just sends a token to yourself? What is the correct way here? I don't want to put it there in cleartext?

The build fails because of [trailing whitespace](https://attrs.visualstudio.com/attrs/_build/results?buildId=33&view=logs&jobId=4902a98b-b56a-5166-3cb6-3d0b3604bae7&taskId=3f5976e1-c73b-5e3a-f2a1-188eab9960a2&lineStart=47&lineEnd=47&colStart=1&colEnd=27) btw.

**Comment by tonybaloney** [comment]
Ah,  that makes more sense if the variable is scoped to only contributors. In that case, you can keep my changes to the codecov job or scrap them. Fixed the trailing whitespace

File: azure-pipelines.yml [hunk 1]
```diff
@@ -7,59 +7,61 @@ jobs:
...
-        condition: succeededOrFailed()
+          versionSpec: '$(python.version)'
+          architecture: 'x64'
+      - script: python -m pip install --upgrade tox codecov
+        displayName: install tox
+
+      - script: tox -e typing
+        displayName: run tox for typing
+
+      - script: tox -e lint,docs
```
Comment (hynek) [inline azure-pipelines.yml]: any reason why you grouped these together?

**Comment by hynek** [comment]
> Ah, that makes more sense if the variable is scoped to only contributors. In that case, you can keep my changes to the codecov job or scrap them. Fixed the trailing whitespace

Yeah but that's kinda bad, because contributors won't get feedback on coverage. :|

I believe I saw something that AP might have their own solution to coverage? Do you know anything about it? Is it ready?

**Comment by tonybaloney** [comment]
I have a solution for coverage gap. Bear with me 🐻

**Comment by hynek** [comment]
giphy

**Comment by tonybaloney** [comment]
@hynek checkout https://attrs.visualstudio.com/attrs/_build/results?buildId=37&view=results and see the summary tab you'll see ~70% coverage, you can download the coverage report ZIP file in the summary page as well, and click on lines to see details.

**Comment by hynek** [comment]
We’ve got 100% coverage though. Is this coverage from only one particular run (maybe just py2 or just py3)?

**Comment by hynek** [comment]
Yeah something is going very wrong:

image

Note the py34, but even that isn't correct. But this might be just us using pytest-cov wrong? I _think_ there was a way to achieve https://hynek.me/articles/testing-packaging/ w/ pytest-cov? 🤔

**Comment by hynek** [comment]
Ah so one think I just noticed: we have special envs for 2.7 and 3.7 because that determines our coverage. The rest of our tox envs used to run w/o coverage for performance reasons (especially PyPy).

With your changes you run the tests that used to _not_ run with coverage with pytest-cov and you've left the coverage-envs using the old way that probably gets ignored?

Is it even possible to combine coverages over multiple runs into AP coverage?

P.S. thanks for all your efforts!

**Comment by hynek** [comment]
thanks for everything, we're on AP now! I think/hope I got all missing things from your PR over time. :)

**Pull request closed without merging**